### PR TITLE
[SG-35217] Wildcard V2: <Typography /> Manual migration `Text`

### PR DESCRIPTION
**Pull request by gitstart-sourcegraph**
## Description
Manual migration required to update our code to use the <Typography /> Wildcard component.
See [Wildcard V2 - Planned work](https://docs.google.com/document/d/1NisbJPiadtt5jQw4vUUYJOr8dD6Urwn5V0mVq2wt6bE/edit#heading=h.g4cw92w3ouhw) for more context

**Note**: Before doing this, we should check if there is scope for a codemod to automate this migration. If so, the codemod should be implemented and ran before starting work on this issue.

## Test Plan
- Ensure there are no UI changes on both `storybook` and `application`

## Ref
[Parent Issue](https://github.com/sourcegraph/sourcegraph/issues/27677)
[SG Issue](https://github.com/sourcegraph/sourcegraph/issues/35217)
[GitStart Issue](https://app.gitstart.com/clients/sourcegraph/tickets/SG-35217)

## Acceptance criteria
 - [ ] All incorrect usage of the previous pattern has been updated to use the Wildcard component
 
### Time estimate
- Pull requests with ~450 lines changed should take 2-3 hours at maximum. Ping the reviewer in the spec pull request if time-consuming changes are required.
- Split the work into multiple pull requests if the total diff is bigger than 450 lines of code.

### Implementation Detail
- Migrate all usage of `P` on codebase to wildcard's `Text`

## App preview:

- [Web](https://sg-web-contractors-sg-35217.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-lavwfdkvqn.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Review by limitedmage — COMMENTED** [review]
@sourcegraph/frontend-platform This seems incredibly extreme to me. What is the justification behind this change? What's wrong with using the `<p>` tag?

**Comment by valerybugakov** [comment]
Hey @limitedmage, here's [the original issue](https://github.com/sourcegraph/sourcegraph/issues/27674#issue-1054782176) with the rationale behind this change:

> ## Why
>  **Unlimited component variations because of the globally styled HTML elements**. Because we rely more on HTML structure than on components in some places (e.g., typography), developers need to keep in mind relative rules applied to certain elements in global styles. With a strict design system, we explicitly defined the limited number of component variations. Now we can move away from the globally styled HTML elements and rely more on strictly styled React components.
> 
> See [the Styling UI guidelines RFC](https://docs.google.com/document/d/15PyMY7cZ-8_UpLFMA66L4DkB2d3Ve37gia2SxLVnE7c/edit#) for more details.

_And more context on the longer-term plan to resurface all pain points that we want to address_

[The Wildcard migrations](https://github.com/sourcegraph/sourcegraph/issues/27606) that we're going through _are not_ the final result we aim for. They are steps in a longer journey described by [the Styling UI guidelines RFC](https://docs.google.com/document/d/15PyMY7cZ-8_UpLFMA66L4DkB2d3Ve37gia2SxLVnE7c/edit#). We want to end up in a codebase with consistent interfaces powered by the Wildcard design system. Apart from fixing inconsistencies in the first place, we want to make it easy to build new interfaces and evolve existing ones in way complaint [with building blocks available in the design system](https://www.figma.com/file/NIsN34NH7lPu04olBzddTw/%E2%9C%B3%EF%B8%8F-Wildcard-Design-System). We already experienced firsthand how hard it was to update the UI of the application without shared/controllable building blocks during [the design refresh](https://github.com/sourcegraph/sourcegraph/issues/19652) one year ago. 
- The first pain point was that global styles unexpectedly affected "random" parts of the application, and it took us a lot longer to update certain parts. [The CSS modules migration](https://github.com/sourcegraph/sourcegraph/issues/23010) solves this problem. ([We're still converting the remaining Bootstrap styles into scoped CSS](https://github.com/sourcegraph/sourcegraph/issues/29200))
- The second paint point was that sometimes the UI implementation differed from feature to feature slightly, which meant that we needed to go through these cases to fix it manually. Having the component interface solves this problem — we can control the API in one place and change it over time. Yes — it's possible to bend the component to your needs via the `className` property and diverge from the design system — but it's more explicit, and defaults are there to guide you.
- The third issue that we bumped into was that in some cases, we need to add to existing "components" `<div className="alert" />` extra markup — icons in case of the alert UI. Thanks to our `browserslist` configuration, we did this by using CSS masks and pseudo-elements. It wouldn't be a problem if we had the `<Alert />` component to incorporate the internal implementation with HTML markup and styles.

To avoid these pitfalls, we need a way to control the interfaces of UI primitives and their internal implementation in one place. 

> What's wrong with using the `<p>` tag?

Coming back to the `Text` component and question about using the `<p>` tag. There's nothing _wrong_ with using it per se, but relying on raw HTML elements, global CSS classes, and global styles don't give us the instruments required to maintain and evolve consistent interfaces. 

The Wildcard design system has [a number of defined `Text` variants](https://www.figma.com/file/NIsN34NH7lPu04olBzddTw/%E2%9C%B3%EF%B8%8F-Wildcard-Design-System?node-id=5601%3A65478) that we want to use across our client application. Introducing the `<Text />` component would help us maintain this mapping by allowing using these defaults via component props. The `Text` migration PR is a first step in introducing it to the codebase. Here we still rely on [global styles defined for all `p` tags](https://github.com/sourcegraph/sourcegraph/blob/main/client/branded/src/global-styles/typography.scss#L2) and on the `<p>` tag `className` prop to produce variants like [`Base/Strong/Body`](https://www.figma.com/file/NIsN34NH7lPu04olBzddTw/%E2%9C%B3%EF%B8%8F-Wildcard-Design-System?node-id=5601%3A65478). In follow-up work these will be deprecated in favor of the `<Text />` component API and the relevant migrations will be completed. E.g. 

```html
<p className="font-weight-medium"><small>content</small></p> 

// becomes

<Typography.Text weight="medium" size="small">content</Text>
```

The API allows us to ban impossible combinations. The one that stands out — [we do not have strong labels](https://www.figma.com/file/NIsN34NH7lPu04olBzddTw/%E2%9C%B3%EF%B8%8F-Wildcard-Design-System?node-id=5601%3A65477) meaning that we can throw a type error in the following example:

```html
<Typography.Label weight="bold" size="regular" />
```

@limitedmage, please, let us know if the context provided helps make sense of the migration initiated in this pull request.

**Comment by limitedmage** [comment]
Does TypeScript imports let just import without having to use the `Typography` namespace? Eg. just use `<Text weight="medium">`. Also, how to we handle inline text changes? Can we nest something like `<Text>Hello <Text weight="medium">world</Text></Text>`?

**Comment by valerybugakov** [comment]
@limitedmage, we're migrating only `<p>` elements for now. For the inline text, we need to do a follow-up migration. We definitely plan to support nesting like you mentioned: `<Text>Hello <Text weight="medium">world</Text></Text>`.

Re: dev experience/readability — at the moment, all the Typography components are exposed only via the namespace, but we can change it. I agree that using namespace all the time might be disturbing. 

Also, @umpox proposed migrating the typography ESLint rule from `react/forbid-elements` to our custom implementation that would provide the auto-fix handler, which would automatically import required components and fix errors on save. I'm adding this idea to our backlog.

File: .eslintrc.js [hunk 1]
```diff
@@ -142,6 +142,10 @@ See https://handbook.sourcegraph.com/community/faq#is-all-of-sourcegraph-open-so
             element: 'h6',
             message: 'Use the <Typography.H6 /> component from @sourcegraph/wildcard instead.',
           },
+          {
+            element: 'p',
+            message: 'Use the <Text /> component from @sourcegraph/wildcard instead.',
```
Comment (valerybugakov) [inline .eslintrc.js:147]:
```suggestion
            message: 'Use the <Text /> component from @sourcegraph/wildcard instead. Check out the RFC for more context: https://bit.ly/3PFw0HM',
```

**Review by valerybugakov — CHANGES_REQUESTED** [review]
A couple of minor changes. Let's address them before merging the PR.

**Review by valerybugakov — APPROVED** [review]
Thanks! LGTM

**Pull request merged**